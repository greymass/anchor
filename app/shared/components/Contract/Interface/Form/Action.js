// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Button, Divider, Form, Header, Message } from 'semantic-ui-react';
import { attempt, isError } from 'lodash';
import { Bytes, Name, Transaction } from '@greymass/eosio';

import { createHttpHandler } from '../../../../utils/http/handler';
import GlobalTransactionModal from '../../../Global/Transaction/Modal';
import GlobalFormFieldGeneric from '../../../Global/Form/Field/Generic';
import WalletModalContentBroadcast from '../../../Wallet/Modal/Content/Broadcast';

const initialState = {
  arrayOptions: {},
  currentArrayValues: {},
  form: {},
  formToggleMatchAccount: {},
};

class ContractInterfaceFormAction extends Component<Props> {
  state = initialState;
  componentDidMount() {
    this.resetForm(this.props.contractAction);
  }
  componentWillReceiveProps(nextProps) {
    const {
      contractAction
    } = this.props;
    if (contractAction !== nextProps.contractAction) {
      this.resetForm(nextProps.contractAction);
    }
  }

  formatField = (contractAction, name, value = '', type = null) => {
    const {
      contract
    } = this.props;
    let fieldType = contract.getFieldType(contractAction, name);

    if (type === 'array') {
      fieldType = 'array';
    }

    switch (fieldType) {
      case 'int': {
        return parseInt(value, 10);
      }
      case 'bool': {
        return value ? 1 : 0;
      }
      case 'array': {
        return value;
      }
      case 'string': {
        return String(value);
      }
      default: {
        if (!isError(attempt(JSON.parse, value))) {
          return JSON.parse(value);
        }
        return String(value);
      }
    }
  };
  onChange = (e, { name, value }, type = null) => {
    const { contractAction } = this.props;
    this.setState({
      form: Object.assign(
        {},
        this.state.form,
        {
          [name]: this.formatField(contractAction, name, value, type)
        }
      )
    });
  };
  onToggle = (e, { name, checked }) => {
    this.setState({
      form: Object.assign(
        {},
        this.state.form,
        {
          [name]: checked ? 1 : 0
        }
      )
    });
  };
  onFormToggleName = (e, { name, checked }) => {
    const { settings } = this.props;
    this.setState({
      form: Object.assign(
        {},
        this.state.form,
        {
          [name]: checked ? settings.account : ''
        }
      ),
      formToggleMatchAccount: Object.assign(
        {},
        this.state.formToggleMatchAccount,
        {
          [name]: checked ? 1 : 0
        }
      )
    });
  };
  onSubmit = async () => {
    const {
      actions,
      connection,
      contract,
      contractAction,
      settings
    } = this.props;
    const { httpClient } = await createHttpHandler(connection);
    const { form } = this.state;
    const fields = contract.getFields(contractAction);
    const modified = Object.assign({}, form);
    for (const field of fields) {
      if (field.type === 'transaction') {
        const requiredAbis = modified[field.name].actions
          .filter((action) =>
            !Bytes.isBytes(action.data) && action.data.constructor.abiName === undefined)
          .map((action) => Name.from(action.account));
        const abis = await Promise.all(requiredAbis.map(async (account) => {
          const result = await httpClient.post(`${connection.httpEndpoint}/v1/chain/get_abi`, {
            account_name: String(account)
          });
          return {
            contract: account,
            abi: result.data.abi
          };
        }));
        const tx = Transaction.from(modified[field.name], abis);
        modified[field.name] = JSON.parse(JSON.stringify(tx));
      }
    }
    actions.buildTransaction(contract, contractAction, settings.account, modified);
  };
  resetForm = (contractAction) => {
    const {
      actions,
      contract
    } = this.props;
    const formData = {};
    const fields = contract.getFields(contractAction);
    fields.forEach((field) => {
      formData[field.name] = this.formatField(contractAction, field.name);
    });
    this.setState({ form: formData });
    actions.clearSystemState();
  };
  render() {
    const {
      actions,
      blockExplorers,
      contract,
      contractAction,
      settings,
      system,
      t,
      transaction,
      unlocked,
    } = this.props;
    const {
      currentArrayValues,
      form,
      formToggleMatchAccount,
    } = this.state;
    const signing = !!(system.TRANSACTION_BUILD === 'PENDING');
    const fields = contract.getFields(contractAction);
    const formFields = [];
    const fieldsMeta = fields.map((field) => {
      if (field.type.substr(field.type.length - 2) === '[]') {
        return { ...field, type: 'array' };
      }
      return field;
    });
    fieldsMeta.forEach((field) => {
      switch (field.type) {
        case 'array': {
          const options = (currentArrayValues[field.name] || [])
            .map(option => ({ text: option, value: option, key: option }))
            .concat([{
              key: '',
              text: t('interface_form_field_options_entry'),
              value: 'placeholder',
              disabled: true
            }]);
          formFields.push((
            <Form.Select
              allowAdditions
              fluid
              label={field.name}
              multiple
              name={field.name}
              options={options}
              search
              selection
              value={currentArrayValues[field.name] || []}
              onChange={(e, { name, value }) => {
                this.setState(
                  { currentArrayValues: { ...this.state.currentArrayValues, [name]: value } },
                  () => {
                    this.onChange(null, { name, value }, field.type);
                  }
                );
              }}
            />
          ));
          break;
        }
        case 'bool': {
          formFields.push((
            <Form.Checkbox
              key={`${contractAction}-${field.name}-${field.type}`}
              label={`${field.name} (${field.type})`}
              name={field.name}
              onChange={this.onToggle}
            />
          ));
          break;
        }
        default: {
          formFields.push((
            <GlobalFormFieldGeneric
              disabled={formToggleMatchAccount[field.name]}
              key={`${contractAction}-${field.name}-${field.type}-${formToggleMatchAccount[field.name]}`}
              label={`${field.name} (${field.type})`}
              name={field.name}
              onChange={this.onChange}
              value={form[field.name]}
            />
          ));
          if (field.type === 'name') {
            formFields.push((
              <Form.Checkbox
                key={`${contractAction}-${field.name}-${field.type}-useaccount`}
                label={`Use current account for field '${field.name}'`}
                name={field.name}
                onChange={this.onFormToggleName}
              />
            ));
          }
        }
      }
    });

    let errors = '';
    if (system.TRANSACTION_BUILD_LAST_ERROR) {
      errors = (
        <Message
          content={String(system.TRANSACTION_BUILD_LAST_ERROR)}
          error
          icon="warning sign"
        />
      );
    }

    let modal;
    if (system.TRANSACTION_SIGN === 'FAILURE'
      && transaction
      && transaction.data
    ) {
      modal = (
        <GlobalTransactionModal
          actionName="TRANSACTION_SIGN"
          actions={actions}
          blockExplorers={blockExplorers}
          content={(
            <WalletModalContentBroadcast
              actions={actions}
              settings={settings}
            />
          )}
          contract={transaction.contract}
          icon="wifi"
          open
          title={t('wallet:wallet_panel_wallet_signbroadcast')}
          settings={settings}
          system={system}
          transaction={transaction}
        />
      );
    } else if (system.TRANSACTION_BROADCAST === 'SUCCESS'
      && transaction
      && transaction.data
    ) {
      modal = (
        <GlobalTransactionModal
          actionName="TRANSACTION_BROADCAST"
          actions={actions}
          blockExplorers={blockExplorers}
          content={(
            <WalletModalContentBroadcast
              actions={actions}
              settings={settings}
            />
          )}
          contract={transaction.contract}
          icon="wifi"
          open
          title={t('wallet:wallet_panel_wallet_signbroadcast')}
          settings={settings}
          system={system}
          transaction={transaction}
        />
      );
    } else if (system.TRANSACTION_BUILD
      && transaction
      && transaction.data
    ) {
      modal = (
        <GlobalTransactionModal
          actionName="TRANSACTION_BUILD"
          actions={actions}
          blockExplorers={blockExplorers}
          content={(
            <WalletModalContentBroadcast
              actions={actions}
              settings={settings}
            />
          )}
          contract={transaction.contract}
          icon="wifi"
          open
          title={t('wallet:wallet_panel_wallet_signbroadcast')}
          settings={settings}
          system={system}
          transaction={transaction}
        />
      );
    }
    return (
      <Form
        error={!!(errors)}
        onSubmit={this.onSubmit}
      >
        <Header>
          {t('interface_form_action_header')}
          <Header.Subheader>
            {t('interface_form_action_subheader')}
          </Header.Subheader>
        </Header>
        <p>
          {t('interface_form_action_instructions')}
        </p>
        <Divider />
        {modal}
        {formFields}
        {errors}
        <Button
          content={t('interface_form_action_button')}
          disabled={signing || !unlocked}
          loading={signing}
          primary
        />
      </Form>
    );
  }
}

export default withTranslation('contract')(ContractInterfaceFormAction);
