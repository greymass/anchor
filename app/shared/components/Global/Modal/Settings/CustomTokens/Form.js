// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Container, Form, Input } from 'semantic-ui-react';

class GlobalModalSettingsCustomTokensForm extends Component<Props> {
  render() {
    const {
      loading,
      onChange,
      onSubmit,
      ref,
      t,
      values
    } = this.props;
    const {
      account,
      symbol,
    } = values;
    return (
      <Form
        loading={loading}
        onSubmit={onSubmit}
        ref={ref}
      >
        <Form.Field
          autoFocus
          control={Input}
          fluid
          label={t('global_modal_settings_customtoken_form_account_label')}
          name="account"
          onChange={onChange}
          placeholder={t('global_modal_settings_customtoken_form_account_placeholder')}
          type="text"
          value={account}
        />
        <Form.Field
          control={Input}
          fluid
          label={t('global_modal_settings_customtoken_form_symbol_label')}
          name="symbol"
          onChange={onChange}
          placeholder={t('global_modal_settings_customtoken_form_symbol_placeholder')}
          type="text"
          value={symbol}
        />
        <Container textAlign="center">
          <Button
            color="green"
            content={t('global_modal_settings_customtoken_add')}
          />
        </Container>
      </Form>
    );
  }
}

export default translate('global')(GlobalModalSettingsCustomTokensForm);
