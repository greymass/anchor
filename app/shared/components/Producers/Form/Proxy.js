// @flow
import React, { Component } from 'react';
import { Button, Divider, Form, Icon, Segment } from 'semantic-ui-react';
import { translate } from 'react-i18next';

import FormFieldGeneric from '../../Global/Form/Field/Generic';
import FormMessageError from '../../Global/Form/Message/Error';
import ProducersFormProxyConfirming from './Proxy/Confirming';

class ProducersFormProxy extends Component<Props> {
  constructor(props) {
    super(props);
    this.state = {
      confirming: false,
      proxyAccountName: ''
    };
  }

  state = {};

  onChange = (e, { name, value }) => {
    const newState = { [name]: value };
    if (name === 'quantity') {
      const [, symbol] = value.split(' ');
      newState.symbol = symbol;
    }
    this.setState(newState);
  }

  onRemove = (e) => {
    this.setState({ proxyAccountName: '' }, () => {
      this.onSubmit();
    });

    e.preventDefault();
    return false;
  }

  onConfirm = () => {
    const {
      proxyAccountName
    } = this.state;

    this.setState({ confirming: false }, () => {
      this.props.actions.voteproducers([], proxyAccountName);
    });
  }

  onSubmit = () => {
    this.setState({
      confirming: true
    });
  }

  onBack = (e) => {
    this.setState({
      confirming: false
    });
    e.preventDefault();
    return false;
  }

  errorsInForm = () => {
    const {
      username
    } = this.state;

    const usernameRegex = /^[a-z]{12}$/;

    if (!usernameRegex.test(username)) {
      return 'not_valid_username';
    }

    return false;
  }

  onError = (error) => {
    let errorMessage;

    if (error !== true) {
      errorMessage = error;
    }

    this.setState({
      submitDisabled: true,
      formError: errorMessage
    });
  }

  render() {
    const {
      onClose,
      system,
      t
    } = this.props;

    const {
      confirming,
      formError,
      proxyAccountName,
      submitDisabled
    } = this.state;

    return (
      <Form
        loading={system.SETPROXY === 'PENDING'}
        onKeyPress={this.onKeyPress}
        onSubmit={this.onSubmit}
      >
        {(confirming)
          ? (
            <ProducersFormProxyConfirming
              onBack={this.onBack}
              onConfirm={this.onConfirm}
              proxyAccountName={proxyAccountName}
            />
          ) : (
            <Segment basic clearing>
              <FormFieldGeneric
                icon="x"
                label={t('producers_form_proxy_label')}
                loading={false}
                name="memo"
                onChange={this.onChange}
                value={proxyAccountName}
              />

              <FormMessageError
                error={formError}
              />
              <Divider />
              <Button
                onClick={onClose}
              >
                <Icon name="x" /> {t('close')}
              </Button>
              <Button
                content={t('producers_form_proxy_confirm')}
                disabled={submitDisabled}
                floated="right"
                onClick={onRemove}
                primary
              />
              <Button
                content={t('producers_form_proxy_remove')}
                disabled={submitDisabled}
                floated="right"
                color="red"
              />
            </Segment>
          )}
      </Form>
    );
  }
}

export default translate('producers')(ProducersFormProxy);
