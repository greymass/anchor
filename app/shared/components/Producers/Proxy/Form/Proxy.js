// @flow
import React, { Component } from 'react';
import { Button, Divider, Form, Icon, Segment, Header, Message } from 'semantic-ui-react';
import { translate } from 'react-i18next';

import ProducersFormProxyConfirming from './Proxy/Confirming';
import GlobalFormFieldAccount from '../../../Global/Form/Field/Account';

class ProducersFormProxy extends Component<Props> {
  constructor(props) {
    super(props);

    this.state = {
      browsingProxies: false,
      confirming: false,
      proxyAccount: '',
      submitDisabled: true
    };
  }

  state = {};

  componentWillMount = () => {
    const {
      addProxy,
      removeProxy
    } = this.props;

    this.setState({
      confirming: addProxy || removeProxy,
      proxyAccount: addProxy
    });
  }

  onChange = (e, { value, valid }) => {
    this.setState({
      submitDisabled: !valid,
      proxyAccount: value
    });
  }

  onKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.onSubmit(e);

      e.preventDefault();
      return false;
    }
  }

  onRemove = (e) => {
    this.setState({ proxyAccount: '' }, () => {
      this.onSubmit();
    });

    e.preventDefault();
    return false;
  }

  onConfirm = () => {
    let {
      proxyAccount
    } = this.state;

    this.setState({ confirming: false }, () => {
      this.props.actions.voteproducers([], proxyAccount || '');
    });
  }

  onSubmit = () => {
    this.setState({
      confirming: true
    });
  }

  onBack = (e) => {
    this.setState({
      confirming: false
    });
    e.preventDefault();
    return false;
  }

  switchFormMode = (e) => {
    this.setState({
      browsingProxies: !this.state.browsingProxies
    });

    e.preventDefault();
    return false;
  }

  render() {
    const {
      addProxy,
      currentProxy,
      isProxying,
      onClose,
      removeProxy,
      settings,
      system,
      t
    } = this.props;

    const {
      confirming,
      proxyAccount,
      submitDisabled
    } = this.state;

    return (
      <Form
        warning
        loading={system.VOTEPRODUCER === 'PENDING'}
        onKeyPress={this.onKeyPress}
        onSubmit={this.onSubmit}
      >
        {(confirming)
          ? (
            <ProducersFormProxyConfirming
              currentProxy={currentProxy}
              showBackButton={!addProxy && !removeProxy}
              onBack={this.onBack}
              onConfirm={this.onConfirm}
              proxyAccount={proxyAccount}
              settings={settings}
            />
          ) : (
            <Segment basic clearing>
              {(isProxying) ? (
                <Header block size="large" style={{ marginTop: 0 }}>
                  <Icon name="circle info" />
                  <Header.Content>
                    <Header.Subheader>
                      {t('producers_table_votes_proxied')}
                    </Header.Subheader>
                    {currentProxy}
                  </Header.Content>
                </Header>
              ) : ''}

              <GlobalFormFieldAccount
                autoFocus
                contacts={settings.contacts}
                label={`${t('producers_form_proxy_label')}:`}
                name="account"
                onChange={this.onChange}
                value={proxyAccount || ''}
              />

              <Message
                content={t('producers_form_proxy_message')}
                warning
              />

              <Divider />
              <Button
                onClick={onClose}
              >
                <Icon name="x" /> {t('close')}
              </Button>
              <Button
                content={t('producers_form_proxy_confirm')}
                disabled={submitDisabled}
                floated="right"
                primary
              />
              {(isProxying) ? (
                <Button
                  content={t('producers_form_proxy_remove')}
                  floated="right"
                  onClick={this.onRemove}
                  color="red"
                />
              ) : ''}
            </Segment>
          )}
      </Form>
    );
  }
}

export default translate('producers')(ProducersFormProxy);
