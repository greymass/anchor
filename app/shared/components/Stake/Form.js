// @flow
import React, { Component } from 'react';
import { I18n } from 'react-i18next';
import { Button, Form, Input, Message, Segment } from 'semantic-ui-react';
import debounce from 'lodash/debounce';

export default class StakeForm extends Component<Props> {
  constructor(props) {
    super(props);

    this.state = {
      error: false,
      cpu_amount: '',
      net_amount: ''
    };

    this.onClick = this.onClick.bind(this);
  }

  onChange = debounce((e, { name, value }) => {
    this.setState({
      error: false,
      [name]: value,
    });
  }, 300)

  onClick = debounce((e, { value }) => {
    const {
      actions,
      account,
      balance
    } = this.props;

    const {
      cpu_amount,
      net_amount
    } = this.state;

    const { setStakeWithValidation } = actions;

    setStakeWithValidation(balance, account, net_amount, cpu_amount)

    this.setState({
      error: false,
      cpu_amount: '',
      net_amount: ''
    });
  }, 300)

  render() {
    const {
      actions,
      validate,
      balance
    } = this.props;

    const {
      clearSettingsCache
    } = actions;

    const {
      error,
      cpu_amount,
      net_amount
    } = this.state;

    return (
      <I18n ns="stake">
        {
          (t) => (
            <Segment basic>
              <Form onSubmit={this.onSubmit}>
                <Form.Field
                  control={Input}
                  fluid
                  label={t('stake_net_amount')}
                  loading={(validate.STAKE === 'PENDING')}
                  name="net_amount"
                  onChange={this.onChange}
                  value={this.state.net_amount}
                />
                <Form.Field
                  control={Input}
                  fluid
                  label={t('stake_cpu_amount')}
                  loading={(validate.STAKE === 'PENDING')}
                  name="cpu_amount"
                  onChange={this.onChange}
                  value={this.state.cpu_amount}
                />
                {(validate.STAKE === 'FAILURE')
                  ? (
                    <Message negative>
                      <p>{t(validate.STAKE_ERROR)}</p>
                    </Message>
                  )
                  : ''
                }
                {(validate.STAKE === 'SUCCESS')
                  ? (
                    <Message positive>
                      <p>{t('stake_updated')}</p>
                    </Message>
                  )
                  : ''
                }
                <Button
                  onClick={this.onClick}
                  content={t('stake_coins')}
                  color="green"
                />
              </Form>
            </Segment>
          )
        }
      </I18n>
    );
  }
}
