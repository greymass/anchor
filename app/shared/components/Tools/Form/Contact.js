// @flow
import React, { Component } from 'react';
import { Segment, Form, Button, Icon } from 'semantic-ui-react';
import { translate } from 'react-i18next';
import { findIndex } from 'lodash';

import GlobalFormFieldAccount from '../../Global/Form/Field/Account';
import GlobalFormFieldGeneric from '../../Global/Form/Field/Generic';
import GlobalFormFieldMemo from '../../Global/Form/Field/Memo';
import GlobalFormMessageError from '../../Global/Form/Message/Error';

class ToolsFormContact extends Component<Props> {
  constructor(props) {
    super(props);
    this.state = {
      accountName: null,
      accountNameValid: true,
      defaultMemo: null,
      defaultMemoValid: true,
      formError: false,
      label: null,
      submitDisabled: true
    };
  }

  componentWillMount() {
    const {
      contactToEdit
    } = this.props;

    if (contactToEdit) {
      const {
        accountName,
        defaultMemo,
        label
      } = contactToEdit;

      this.setState({
        accountName,
        defaultMemo,
        label,
      });
    }
  }

  onChange = (e, { name, value, valid }) => {
    this.setState({
      formError: null,
      submitDisabled: false,
      [`${name}Valid`]: valid,
      [name]: value
    }, () => {
      const error = this.errorInForm();

      if (error) {
        let formError;

        if (error !== true) {
          formError = error;
        }
        this.setState({
          formError,
          submitDisabled: true
        });
      }
    });
  }

  errorInForm = () => {
    const {
      contacts,
      contactToEdit,
    } = this.props;

    const {
      accountName,
      accountNameValid,
      defaultMemoValid
    } = this.state;

    if (!defaultMemoValid) {
      return 'invalid_memo';
    }

    if (!accountName || accountName.length === 0) {
      return true;
    }

    if (!accountNameValid) {
      return 'invalid_accountName';
    }

    const accountNameHasChanged = !contactToEdit || contactToEdit.accountName !== accountName;
    const accountNameIsInList = findIndex(contacts, { accountName }) > -1;

    if (accountNameHasChanged && accountNameIsInList) {
      return 'accountName_not_unique_in_contacts';
    }

    return false;
  }

  onSubmit = () => {
    const {
      actions,
      contactToEdit,
      contacts,
      deleteContact,
      onSuccess
    } = this.props;

    const {
      accountName,
      defaultMemo,
      label
    } = this.state;

    if (contactToEdit) {
      deleteContact(contactToEdit);
    }

    actions.setSetting(
      'contacts',
      contacts.concat({
        accountName,
        defaultMemo,
        label
      })
    );

    this.setState({
      accountName: null,
      defaultMemo: null,
      label: null
    }, () => {
      onSuccess((contactToEdit) ? 'tools_contacts_success_edit' : null);
    });
  }

  render() {
    const {
      contacts,
      onClose,
      t
    } = this.props;

    const {
      accountName,
      defaultMemo,
      formError,
      label,
      submitDisabled
    } = this.state;

    return (
      <Form
        onKeyPress={this.onKeyPress}
        onSubmit={this.onSubmit}
      >
        <GlobalFormFieldAccount
          contacts={contacts}
          label={t('tools_form_contact_account_name')}
          name="accountName"
          onChange={this.onChange}
          value={accountName || ''}
        />
        <GlobalFormFieldGeneric
          value={label || ''}
          label={t('tools_form_contact_label')}
          name="label"
          onChange={this.onChange}
        />
        <GlobalFormFieldMemo
          value={defaultMemo || ''}
          label={t('tools_form_contact_default_memo')}
          name="defaultMemo"
          onChange={this.onChange}
        />

        <GlobalFormMessageError
          error={formError}
          icon="warning sign"
        />

        <Segment basic clearing>
          <Button
            content={t('tools_form_contact_submit')}
            color="green"
            disabled={submitDisabled}
            floated="right"
            primary
          />
          <Button
            onClick={onClose}
          >
            <Icon name="x" /> {t('tools_form_contact_cancel')}
          </Button>
        </Segment>
      </Form>
    );
  }
}

export default translate('tools')(ToolsFormContact);
