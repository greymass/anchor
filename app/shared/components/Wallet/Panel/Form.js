// @flow
import React, { Component } from 'react';
import { I18n } from 'react-i18next';
import { Divider, Form } from 'semantic-ui-react';
import debounce from 'lodash/debounce';

import WalletPanelFormAccount from './Form/Account';
import WalletPanelFormEncrypt from './Form/Encrypt';
import WalletPanelFormError from './Form/Error';
import WalletPanelFormKey from './Form/Key';
import WalletPanelFormNode from './Form/Node';
import WalletPanelFormSubmit from './Form/Submit';

export default class WalletPanelForm extends Component<Props> {
  constructor(props) {
    super(props);
    this.state = {
      confirming: false,
      formData: Object.assign({}, {
        encryptWallet: true,
        password: ''
      }, props.keys, props.settings)
    };
  }

  componentDidMount() {
    // Validate settings on app start
    const {
      actions
    } = this.props;
    const {
      formData
    } = this.state;
    actions.validateKey(formData.key);
  }

  onChange = debounce((e, { name, value }) => {
    this.setState({
      formData: { ...this.state.formData, [name]: value },
    });
    // Determine if field should be saved in persistent settings
    if (['account', 'node'].indexOf(name) > -1) {
      const {
        actions
      } = this.props;
      const { setSettingWithValidation } = actions;
      setSettingWithValidation(name, value);
    }
  }, 300)

  onKeyChange = debounce((e, { name, value }) => {
    const {
      actions
    } = this.props;
    const { validateKey } = actions;
    validateKey(value);
    this.setState({
      formData: { ...this.state.formData, [name]: value },
    });
  }, 300)

  onToggle = (e, { name, checked }) => {
    this.setState({
      formData: { ...this.state.formData, [name]: (checked) },
    });
  }

  onConfirm = () => this.setState({ confirming: true });

  onCancel = () => this.setState({ confirming: false });

  onSubmit = () => {
    const {
      actions
    } = this.props;
    const {
      formData
    } = this.state;
    const {
      encryptWallet,
      key,
      password
    } = formData;
    const {
      setWalletKey,
      setTemporaryKey
    } = actions;
    if (encryptWallet) {
      setWalletKey(key, password);
    } else {
      setTemporaryKey(key);
    }
    this.setState({
      confirming: false
    });
  }

  render() {
    const {
      validate
    } = this.props;
    const {
      confirming,
      formData,
      error
    } = this.state;
    const {
      account,
      encryptWallet,
      key,
      node,
      password
    } = formData;
    const validUser = (
      validate.NODE === 'SUCCESS'
      && validate.ACCOUNT === 'SUCCESS'
      && validate.KEY === 'SUCCESS'
      && (!encryptWallet || (encryptWallet && password.length > 0))
    );
    return (
      <I18n ns="wallet">
        {
          (t) => (
            <Form
              onSubmit={
                (encryptWallet)
                  ? this.onConfirm
                  : this.onSubmit
              }
            >
              <WalletPanelFormNode
                onChange={this.onChange}
                validate={validate}
                value={node}
              />
              <WalletPanelFormAccount
                onChange={this.onChange}
                validate={validate}
                value={account}
              />
              <WalletPanelFormKey
                onChange={this.onKeyChange}
                validate={validate}
                value={key}
              />
              <Divider />
              <WalletPanelFormEncrypt
                encryptWallet={encryptWallet}
                onChange={this.onChange}
                onToggle={this.onToggle}
              />
              <Divider />
              <WalletPanelFormError
                error={t(error)}
              />
              <WalletPanelFormSubmit
                confirming={confirming}
                encryptWallet={encryptWallet}
                onCancel={this.onCancel}
                onConfirm={this.onConfirm}
                onSubmit={this.onSubmit.bind(this)}
                password={password}
                validUser={validUser}
              />
            </Form>
          )
        }
      </I18n>
    );
  }
}
