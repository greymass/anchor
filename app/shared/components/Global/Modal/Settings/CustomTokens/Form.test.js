import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalModalSettingsCustomTokensForm } from './Form';

const mockProps = {
  onSubmit: jest.fn,
  t: text => text,
  token: { issuer: 'teamgreymass' },
  values: { account: 'eosio', symbol: 'EOS'},
};

describe('GlobalModalSettingsCustomTokensForm', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalModalSettingsCustomTokensForm {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
