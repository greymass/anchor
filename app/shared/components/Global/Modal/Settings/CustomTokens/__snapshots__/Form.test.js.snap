// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalModalSettingsCustomTokensForm renders correctly 1`] = `
<Form
  as="form"
  onSubmit={[Function]}
>
  <FormField
    autoFocus={true}
    control={[Function]}
    fluid={true}
    label="global_modal_settings_customtoken_form_account_label"
    name="account"
    placeholder="global_modal_settings_customtoken_form_account_placeholder"
    type="text"
    value="eosio"
  />
  <FormField
    control={[Function]}
    fluid={true}
    label="global_modal_settings_customtoken_form_symbol_label"
    name="symbol"
    placeholder="global_modal_settings_customtoken_form_symbol_placeholder"
    type="text"
    value="EOS"
  />
  <Container
    textAlign="center"
  >
    <Button
      as="button"
      color="green"
      content="global_modal_settings_customtoken_add"
    />
  </Container>
</Form>
`;
